homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,320 @@
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, self._icon)
+
+            self._state = new_state.state
+
+            if self._state in [STATE_UNKNOWN, STATE_UNAVAILABLE]:
+                return
+
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s, %s) -> %s", filt.name,
+                                  self._entity,
+                                  self._state,
+                                  "skip" if filt.skip_processing else
+                                  filtered_state)
+                    if filt.skip_processing:
+                        return
+                except ValueError:
+                    _LOGGER.warning("Could not convert state: %s to number",
+                                    self._state)
+                finally:
+                    self._state = filtered_state
+
+            self.async_schedule_update_ha_state()
+
+        async_track_state_change(
+            self.hass, self._entity, filter_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        state_attr = {
+            ATTR_ENTITY_ID: self._entity
+        }
+        for filt in self._filters:
+            for filt_stat_key, filt_stat_value in filt.stats.items():
+                filt_stat = "{}_{}".format(filt.name, filt_stat_key)
+                _LOGGER.debug("stats(%s): %s: %s", self._entity,
+                              filt_stat, filt_stat_value)
+                state_attr.update({
```

Comment:
This is so expensive and inefficient. Create a dict with 1 value to then pass to the update method after which it is discarded? 🤔  You can do `state_atr[filt_stat] = filt_stat_value` but really, just go for the define it at once declaration I suggested earlier or explain to me why it is so important to log every attribute individually?

Reply:
@robmarkcole and @OttoWinter in the previous PR made a case for having stats of the filters. 

Personally I think logging is enough to tune the filters, logging can even be done in the filter itself saving the dict. 

If I don't get feedback from you guys, I'm taking the stats off from this PR.

Reply:
I say leave the stats of for now, since I coudn't even say what stats I'd request. Once I've been using the component a while then will be in a position to do some analysis and make a suggestion, cheers

Reply:
I don't think we should do stat reporting through attributes (if at all). I personally hate it when entities put completely unnecessary stuff in their attributes that just clutter the front end (and state machine)

Reply:
Then it's settled! :) uff ... one less feature to creep this PR

Reply:
Databases all over the world are cheering 😉